### PR TITLE
feat(tools): create ui-components package and setup Storybook

### PR DESCRIPTION
**Pull request by huyenltnguyen**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

## Description

Ref #41580 

This PR sets up Storybook in the `client` folder. I try to keep the PR small with some basic configs, and will have follow-up PRs to iterate. Storybook right now has a stub component that was generated by Storybook's installation script. 

To start it, run:

```
npm run storybook
```

---

I will add comments to the code to give a better explanation, but here is a breakdown of what I did:
- I followed the [Installation doc](https://storybook.js.org/docs/react/get-started/install) and ran `npx sb init` inside `/client`. The script creates a `.storybook` folder right under `/client`, and a `stories` folder under `/client/src`
- The components files were actually created in TypeScript, I deleted most of them, kept the `Button` component only and converted it to JavaScript. The component list that the script generates can be found [here](https://github.com/storybookjs/storybook/tree/next/lib/cli/src/frameworks/react/ts), and the generated introduction file can be found [here](https://github.com/storybookjs/storybook/blob/next/lib/cli/src/frameworks/common/Introduction.stories.mdx)
- Installed `@storybook/addon-postcss` as a dev dependency and added it to the `.storybook/main.js` config
- Installed `dotenv-webpack` and `html-webpack-plugin` as dev dependencies as a workaround for https://github.com/storybookjs/storybook/issues/14403 (I will elaborate more in the comment)
- Themed Storybook a little by adding freeCodeCamp logo
- Added the storybook command to the root `package.json` file

## Screenshot

Screen Shot 2021-04-25 at 02 01 47

### CODE REVIEW

File: client/.storybook/main.js
Comment (huyenltnguyen) [inline client/.storybook/main.js:6]: I got the following message when running Storybook without this addon. But to be honest, I'm not sure if we need it.

Screen Shot 2021-04-25 at 02 09 35

File: client/.storybook/manager.js
Comment (huyenltnguyen) [inline client/.storybook/manager.js:1]: This file wasn't generated by the script. I created it as part of the custom theming.

Ref: https://storybook.js.org/docs/react/configure/theming#create-a-theme-quickstart

File: client/.storybook/theme.js
Comment (huyenltnguyen) [inline client/.storybook/theme.js:1]: This file wasn't generated by the script. I created it as part of the custom theming.

Ref: https://storybook.js.org/docs/react/configure/theming#create-a-theme-quickstart

File: client/.storybook/theme.js
Comment (huyenltnguyen) [inline client/.storybook/theme.js:8]: I browsed the cdn repo but couldn't find a logo file that has both text and gray90 background color. I resorted to the secondary logo as the text of the primary one is not readable on Storybook's default sidebar background.

File: client/package.json
Comment (huyenltnguyen) [inline client/package.json:40]: These were added by the Storybook script.

File: client/package.json
Comment (huyenltnguyen) [inline client/package.json:135]: All of the new dependencies were added automatically by Storybook, except for `@storybook/addon-postcss`.

File: client/package.json
Comment (huyenltnguyen) [inline client/package.json:140]: If I understand https://github.com/storybookjs/storybook/issues/14403 correctly, Storybook is using a very old version of these libraries. They are hoisted and got picked up by `webpack`, but we got error because they are not compatible with webpack 5. 

Apparently Storybook team fixed it but then reverted the PR due to it being backward incompatible (https://github.com/storybookjs/storybook/issues/14403#issuecomment-817138766). I followed https://github.com/storybookjs/storybook/issues/14497#issuecomment-818838975 and added these packages to our dev dependency list as a workaround.

File: client/src/stories/Introduction.stories.mdx
Comment (huyenltnguyen) [inline client/src/stories/Introduction.stories.mdx:2]: This file is generated by Storybook, along with a bunch of svg files. It has links to the official document pages, which I think might come in handy especially when we don't have documentation on our side yet.

I decided to keep this one, and we can delete it once our component library is more mature.

File: client/package.json
Comment (huyenltnguyen) [reply client/package.json:140]: The Storybook team recommends using Yarn's `resolutions` to workaround this issue. It is not applicable to us, but I'm just posting the link here for future reference: https://gist.github.com/shilman/8856ea1786dcd247139b47b270912324#yarn-resolutions

**Review by raisedadead — COMMENTED** [review]
Nice work @huyenltnguyen!

I just wanted to drop in and give some quick feedback about the structure. I think the best place to develop and design the components would be within the [`tools`](https://github.com/freeCodeCamp/freeCodeCamp/tree/main/tools) location. We can add additional directory within the tools called `ui-components` and name the package `@freeCodeCamp/ui` - not sure we will publish it though.

I say this because we may want to extend the components well beyond their use in the client app (which is architecturally speaking a consumer).

The ability to build components in isolation would force to think of them as generic components vs customizing too much.

IMHO, if we use [tailwind ](https://tailwindcss.com/) with [headless-ui](https://headlessui.dev/) I am convinced 90% of our needs will be addressed already.

With storybook we can address the custom needs like layouts, nav and theming.

Thinking ahead when we have additional apps (like a JAMStack news) this could be the common locations to consume components from.

P.S: This will also keep the client app lean.

**Comment by huyenltnguyen** [comment]
@raisedadead Totally agreed! I'm also in favor of keeping the components platform-agnostic, so it makes sense to move it out of `client`. I was under the impression that we would built the components from scratch, but `headless-ui` looks promising. 

I'll try to get back to this PR over the weekend 😄  

In the meantime, I'm wondering if you have any suggestions on the folder structure. Storybook's default structure is quite flat:

```
stories
├── Button.css
├── Button.js
├── Button.stories.js
```

I'm thinking about having a folder for each component, which would be consistent with our `components` folder in `client`, and an `index.js` file that serves as an entry point for component export. It should be something like this:

```
ui-components
├── Button
│   ├── Button.css
│   ├── Button.js
│   ├── Button.stories.js
│   ├── Button.test.js
├── index.js
```
---

One thing I completely forgot in the issue discussion is that we should write unit tests for our components as well. 

For components that `headless-ui` doesn't support yet and we need to build ourselves, we should try to ensure accessibility requirements are covered via unit tests. An example could be making sure icon-only buttons have `aria-label` specified, or disabled buttons have the `disabled` attribute set, etc.. 

I have some experience working with Jest and [@testing-library/react](https://testing-library.com/docs/react-testing-library/intro/) and I guess I'm going to propose that 😛 But I would like to get some input from the core team on the testing approach as well.

**Comment by raisedadead** [comment]
I like the idea of having a simple flat structure instead of layers like we currently do. One thing to note is that all file names should be kebab-case

As for tests, YESSSS PLEASE!!! we are fans of RTL so that should already align with what we want. Please go ahead.

**Review by huyenltnguyen — COMMENTED** [review]
Okay, the PR is ready for another review round.

I reverted the changes to start clean, then:
- cd into `tools` and ran `npx sb init --type react`. The command is from the Troubleshooting section in the Storybook [installation guide](https://storybook.js.org/docs/react/get-started/install)
-  The `package.json` file generated from the script didn't not include React. I believe this is because Storybook is meant to piggyback a project with a framework installed. I added `react` and `react-dom` to the dependencies list to allow Storybook to run
- The rest are the same as my initial setup

What I think I'm missing here is the setup for package publication. I will need to read up on this as I have never released a package before.

I will leave Tailwind and RTL setup for another PR. They are already in the TODO list https://github.com/freeCodeCamp/freeCodeCamp/issues/41580#issuecomment-817223547 🙂

**Comment by ahmadabdolsaheb** [comment]
> I like the idea of having a simple flat structure instead of layers like we currently do. One thing to note is that all file names should be kebab-case

I also like having a simple structure for the components that share the same hierarchy. Maybe we should consider adding another directory when the components are not atomic and are a combination of UI elements such as forms, navigation, etc.

File: client/.storybook/theme.js
Comment (ahmadabdolsaheb) [reply client/.storybook/theme.js]: We have png assets with set backgrounds on the design style guide. https://github.com/freeCodeCamp/design-style-guide/blob/master/downloads/fcc_primary_large.png but they are not on the CDN.

We could default to the gray85 for the background and use the primary SVG down the road.

File: client/.storybook/main.js
Comment (ahmadabdolsaheb) [reply client/.storybook/main.js]: Other than the error, we would probably need it for tailwind.

**Review by raisedadead — APPROVED** [review]
Going to go ahead merge, and unblock you from spiking further.

**Comment by raisedadead** [comment]
image


Is is possible to edit this intro view?

File: client/.storybook/theme.js
Comment (huyenltnguyen) [reply client/.storybook/theme.js]: I just tried it out and the PNG doesn't seem to have a background either

Screen Shot 2021-05-11 at 14 05 12

---
Storybook has very limited [theming options](https://github.com/storybookjs/storybook/blob/a3d63d6a3fe28b6e3577ba9e50370a0d14ceca7e/lib/theming/src/types.ts#L5). In order to have a dark background for the primary logo, I would need to set `gray85` to the entire sidebar and change the text color; but it affects the Controls panel, which doesn't have dedicated styling options.

So I think we would still need an image file with a set background on CDN if we want to use the primary logo. 

Screen Shot 2021-05-11 at 13 50 49

File: client/.storybook/main.js
Comment (thecodingaviator) [reply client/.storybook/main.js]: Without addon-postcss, the code will throw a deprecation warning. It is needed

**Pull request merged**
